### PR TITLE
NIAD-3144 - Confidentiality Code (i.e. Redactions) Support for `SpecimenMapper` and `DiagnosticReportMapper`

### PR DESCRIPTION
**Pull request by martin-nhs**
## What

This is the first of two pull requests aimed at introducing support for confidentiality codes in lab results. In this PR, support is added to the DiagnosticReportMapper and SpecimenMapper classes, enabling them to handle confidentiality codes effectively.

## Why

More details can be found [on the ticket](https://nhse-dsic.atlassian.net/browse/NIAD-3144).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated the [Changelog](CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 1]
```diff
@@ -87,7 +89,8 @@ public String mapDiagnosticReportToCompoundStatement(DiagnosticReport diagnostic
             .extensionId(fetchExtensionId(diagnosticReport.getIdentifier()))
             .availabilityTimeElement(StatementTimeMappingUtils.prepareAvailabilityTime(diagnosticReport.getIssuedElement()))
             .narrativeStatements(reportLevelNarrativeStatements)
-            .specimens(mappedSpecimens);
+            .specimens(mappedSpecimens)
+            .confidentialityCode(confidentialityService.generateConfidentialityCode(diagnosticReport).orElse(null));
```
Comment (martin-nhs) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:93]: Bring this out onto a new line please

File: service/src/test/java/uk/nhs/adaptors/gp2gp/TestUtility.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            displayName="no disclosure to patient, family or caregivers without attending provider's authorization" />
+        """;
+
+    public static void assertThatXmlContainsNopatConfidentialityCode(String xml) {
+        assertThat(xml).contains(NOPAT_CONFIDENTIALITY_CODE);
+    }
+
+    public static void assertThatXmlDoesNotContainConfidentialityCode(String xml) {
+        assertThat(xml).doesNotContain(CONFIDENTIALITY_CODE);
+    }
```
Comment (martin-nhs) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/TestUtility.java:29]: Are these even required? Probably not!

Comment (martin-nhs) [reply service/src/test/java/uk/nhs/adaptors/gp2gp/TestUtility.java]: Removed.

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java [hunk 3]
```diff
@@ -118,15 +133,52 @@ public void tearDown() {
...
+        );
+    }
+
+    @Test
+    void When_DiagnosticReport_With_NoscrubMetaSecurity_Expect_ConfidentialityCodeNotWithinCompoundStatement() {
+        final String testFile = "diagnostic-report-with-multi-specimens-noscrub.json";
+        final DiagnosticReport diagnosticReport = getDiagnosticReportResourceFromJson(testFile);
+
+        when(confidentialityService.generateConfidentialityCode(diagnosticReportCaptor.capture()))
+            .thenReturn(Optional.empty());
```
Comment (martin-nhs) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapperTest.java:167]: The captor isn't required here, we can just pass the resource itself.

```suggestion
        when(confidentialityService.generateConfidentialityCode(diagnosticReport))
            .thenReturn(Optional.empty());
```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java [hunk 4]
```diff
@@ -87,7 +89,8 @@ public String mapDiagnosticReportToCompoundStatement(DiagnosticReport diagnostic
             .extensionId(fetchExtensionId(diagnosticReport.getIdentifier()))
             .availabilityTimeElement(StatementTimeMappingUtils.prepareAvailabilityTime(diagnosticReport.getIssuedElement()))
             .narrativeStatements(reportLevelNarrativeStatements)
-            .specimens(mappedSpecimens);
+            .specimens(mappedSpecimens)
```
Comment (ole4ryb) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:92]: I think the whole section (which confidentialityCode is part of) can be extracted into a separate method with the same name diagnosticReportCompoundStatementTemplateParameters.

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java [hunk 5]
```diff
@@ -61,17 +62,21 @@ public class SpecimenMapper {
     private final MessageContext messageContext;
     private final ObservationMapper observationMapper;
     private final RandomIdGeneratorService randomIdGeneratorService;
+    private final ConfidentialityService confidentialityService;
 
     public String mapSpecimenToCompoundStatement(Specimen specimen, List<Observation> observations, DiagnosticReport diagnosticReport) {
         String availabilityTimeElement = StatementTimeMappingUtils.prepareAvailabilityTime(diagnosticReport.getIssuedElement());
         String mappedObservations = mapObservationsAssociatedWithSpecimen(specimen, observations);
 
+        final Optional<String> confidentialityCode = confidentialityService.generateConfidentialityCode(specimen);
```
Comment (ole4ryb) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/SpecimenMapper.java:71]: I think we can move this line into line 79 and remove confidentialityCode variable similarly as it is done in DiagnosticReportMapper

File: service/src/test/java/uk/nhs/adaptors/gp2gp/utils/ConfidentialityCodeUtility.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    private static Coding getNoscrubCoding() {
+        final String code = "NOSCRUB";
+        final String display = "no scrubbing of the patient, family or caregivers without attending provider's authorization";
+
+        return getCoding(code, display);
+    }
+
+    private static Coding getCoding(String code, String display) {
+        final Coding coding = new Coding();
```
Comment (ole4ryb) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/utils/ConfidentialityCodeUtility.java:62]: I think you can use FluentAPI here, less code the same result:
final Coding coding = new Coding()
                                    .setCode(code)
                                    .setSystem("http://hl7.org/fhir/v3/ActCode")
                                    .setDisplay(display);
        return coding;

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java (hunk 4 above)
Comment (martin-nhs) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/diagnosticreport/DiagnosticReportMapper.java:92]: TODO separate clean-up PR, out of scope for NIAD-3144

**Pull request merged**